pkg/controller/blockdevice/effects.go

```diff
@@ -0,0 +1,423 @@
+		"device":  bd.Name,
+		"context": "Effect",
+	})
+}
+
+// jitterPoll polls the conditionFunc with time jittering.
+// The conditionFunc here has a slight different semantic: if `done` is true,
+// no matter there is an error with it or not, jitterPoll always ceases.
+func jitterPoll(e effectController, bd *diskv1.BlockDevice, conditionFunc wait.ConditionFunc) {
+	stopCh := make(chan struct{})
+	doneCh := make(chan struct{})
+	currentPhase := bd.Status.ProvisionPhase
+	go func() {
+		wait.JitterUntil(func() {
+			done, err := conditionFunc()
+			if done {
+				if err != nil {
+					logEffect(bd).Errorf("Ceased phase %s with err: %v", currentPhase, err.Error())
+				} else {
+					logEffect(bd).Infof("Finished phase %s", currentPhase)
+				}
+				stopCh <- struct{}{}
+			} else {
+				if err != nil {
+					logEffect(bd).Errorf("Failed during phase %s: %v", currentPhase, err.Error())
+				} else {
+					logEffect(bd).Infof("Continue polling in phase %s", currentPhase)
+				}
+			}
+		}, effectDefaultPollInterval, 1.0, false, stopCh)
+		doneCh <- struct{}{}
+	}()
+
+	go func() {
+		select {
+		case <-time.After(effectDefaultTimeout):
+			// Stop JitterUntil
+			stopCh <- struct{}{}
+			onEffectTimeout(e, bd)
+		case <-doneCh:
+			logEffect(bd).Debugf("Invalidated timeout: finish %s in time", currentPhase)
+		}
+	}()
+}
+
+func onEffectTimeout(e effectController, bd *diskv1.BlockDevice) {
+	err := fmt.Errorf("Timeout in phase %s", bd.Status.ProvisionPhase)
+	logEffect(bd).Error(err)
+	newBd, getErr := e.BlockdeviceCache().Get(bd.Namespace, bd.Name)
+	if getErr != nil {
+		logEffect(bd).Errorf("Failed to update after timeout in phase %s: %v", bd.Status.ProvisionPhase, getErr.Error())
+		return
+	}
+	newBdCpy := newBd.DeepCopy()
+	setDeviceFailedCondition(newBdCpy, corev1.ConditionTrue, err.Error())
+	diskv1.ProvisionPhaseFailed.Set(newBdCpy)
+	if _, err := e.Blockdevices().Update(newBdCpy); err != nil {
+		logEffect(bd).Errorf("Failed to update after timeout in phase %s: %v", bd.Status.ProvisionPhase, err.Error())
+	}
+}
+
+type cmdResultUpdater func(*diskv1.BlockDevice) *diskv1.BlockDevice
+
+func onCmdTimeout(e effectController, bd *diskv1.BlockDevice, f func(chan<- cmdResultUpdater)) {
+	doneCh := make(chan cmdResultUpdater, 1)
+	go func() {
+		select {
+		case <-time.After(effectDefaultTimeout):
+			onEffectTimeout(e, bd)
+		case update := <-doneCh:
+			if _, err := e.Blockdevices().Update(update(bd.DeepCopy())); err != nil {
+				emitError := func(err error) {
+					logEffect(bd).Errorf("Failed to update in phase %s: %v", bd.Status.ProvisionPhase, err.Error())
+				}
+				if !errors.IsConflict(err) {
+					emitError(err)
+					return
+				}
+				newBd, err := e.BlockdeviceCache().Get(bd.Namespace, bd.Name)
+				if err != nil {
+					emitError(err)
+					return
+				}
+				if _, err := e.Blockdevices().Update(update(newBd.DeepCopy())); err != nil {
+					emitError(err)
+					return
+				}
+			}
+		}
+	}()
+	go func() { f(doneCh) }()
```

Comment:
I suppose this would be sufficient:
```suggestion
	go f(doneCh)
```

Reply:
fixed 30893ee